new-packages/ts-project/__tests__/state.test.ts

```diff
@@ -0,0 +1,695 @@
+            "data": {
+              "description": undefined,
+              "entry": [],
+              "exit": [],
+              "history": undefined,
+              "initial": "foo",
+              "invoke": [],
+              "metaEntries": [],
+              "tags": [],
+              "type": undefined,
+            },
+            "id": "",
+            "states": [
+              {
+                "data": {
+                  "description": undefined,
+                  "entry": [],
+                  "exit": [],
+                  "history": undefined,
+                  "initial": undefined,
+                  "invoke": [],
+                  "metaEntries": [],
+                  "tags": [],
+                  "type": undefined,
+                },
+                "id": "foo",
+                "states": [],
+              },
+            ],
+          },
+          "states": [],
+        },
+        [],
+      ],
+    ]
+  `);
+});
+
+test('Should extract state.initial with undefined value', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        initial: undefined,
+        states: {},
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  expect(project.extractMachines('index.ts')).toMatchInlineSnapshot(`
+    [
+      [
+        {
+          "edges": [],
+          "rootState": {
+            "data": {
+              "description": undefined,
+              "entry": [],
+              "exit": [],
+              "history": undefined,
+              "initial": undefined,
+              "invoke": [],
+              "metaEntries": [],
+              "tags": [],
+              "type": undefined,
+            },
+            "id": "",
+            "states": [],
+          },
+          "states": [],
+        },
+        [],
+      ],
+    ]
+  `);
+});
+
+test('Should raise error when state.initial property has invalid value', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        initial: {},
+        states: {},
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  expect(project.extractMachines('index.ts')).toMatchInlineSnapshot(`
+    [
+      [
+        {
+          "edges": [],
+          "rootState": {
+            "data": {
+              "description": undefined,
+              "entry": [],
+              "exit": [],
+              "history": undefined,
+              "initial": undefined,
+              "invoke": [],
+              "metaEntries": [],
+              "tags": [],
+              "type": undefined,
+            },
+            "id": "",
+            "states": [],
+          },
+          "states": [],
+        },
+        [
+          {
+            "node": NodeObject {
```

Comment:
We should remove `node` from errors in a followup PR. It's super noisy and we might not want to use this directly anyhow. I think it's better to attach custom metadata / text spans to those error objects. We don't have to spend time on this at all right now though. It's enough to recognize the errors for now and we'll take care of being able to actually display them later on.